netann/host_ann.go

```diff
@@ -0,0 +1,133 @@
+
+	// LookupHost performs DNS resolution on a given host and returns its
+	// addresses.
+	LookupHost func(string) (net.Addr, error)
+
+	// AnnounceNewIPs announces a new set of IP addresses for the backing
+	// Lightning node. The first set of addresses is the new set of
+	// addresses that we should advertise, while the other set are the
+	// stale addresses that we should no longer advertise.
+	AnnounceNewIPs func([]net.Addr, map[string]net.Addr) error
+}
+
+// HostAnnouncer is a sub-system that allows a user to specify a set of hosts
+// for lnd that will be continually resolved to notice any IP address changes.
+// If the target IP address for a host changes, then we'll generate a new
+// NodeAnnouncement that includes these new IPs.
+type HostAnnouncer struct {
+	cfg HostAnnouncerConfig
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewHostAnnouncer returns a new instance of the HostAnnouncer.
+func NewHostAnnouncer(cfg HostAnnouncerConfig) *HostAnnouncer {
+	return &HostAnnouncer{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}
+}
+
+// Start starts the HostAnnouncer.
+func (h *HostAnnouncer) Start() error {
+	h.wg.Add(1)
+	go h.hostWatcher()
+
+	return nil
+}
+
+// Stop signals the HostAnnouncer for a graceful stop.
+func (h *HostAnnouncer) Stop() error {
+	close(h.quit)
+	h.wg.Wait()
+
+	return nil
+}
+
+// hostWatcher periodically attempts to resolve the IP for each host, updating
+// them if they change within the interval.
+func (h *HostAnnouncer) hostWatcher() {
+	defer h.wg.Done()
+
+	ipMapping := make(map[string]net.Addr)
+	refreshHosts := func() {
+
+		// We'll now run through each of our hosts to check if they had
+		// their backing IPs changed. If so, we'll want to re-announce
+		// them.
+		var addrsToUpdate []net.Addr
+		addrsToRemove := make(map[string]net.Addr)
+		for _, host := range h.cfg.Hosts {
+			newAddr, err := h.cfg.LookupHost(host)
+			if err != nil {
+				log.Warnf("unable to resolve IP for "+
+					"host %v: %v", host, err)
+				continue
+			}
+
+			oldAddr, ok := ipMapping[host]
+			if ok && oldAddr.String() == newAddr.String() {
+				continue
+			}
+
+			log.Debugf("IP change detected! %v: %v -> %v", host,
+				oldAddr, newAddr)
+
+			ipMapping[host] = newAddr
+
+			// If we had already advertised an addr for this host,
+			// then we'll need to remove that old stale address.
+			if oldAddr != nil {
+				addrsToRemove[oldAddr.String()] = oldAddr
+			}
+
+			addrsToUpdate = append(addrsToUpdate, newAddr)
+		}
+
+		// If we don't have any addresses to update, then we can skip
+		// things around until the next round.
+		if len(addrsToUpdate) == 0 {
+			return
+		}
+
+		// Now that we know the set of IPs we need to update, we'll do
+		// them all in a single batch.
+		err := h.cfg.AnnounceNewIPs(addrsToUpdate, addrsToRemove)
+		if err != nil {
+			log.Warnf("unable to announce new IPs: %v", err)
+		}
+	}
+
+	refreshHosts()
+
+	for {
+		select {
+		case <-h.cfg.RefreshTicker.Ticks():
```

Comment:
ticker is never started

Reply:
Ah good catch, I'm used to the one in the stdlib that is started once you create it.